src/thunderbots/software/ai/world/team.cpp

```diff
@@ -1,33 +1,145 @@
+    {
+        auto duplicate_id = robot_ids.insert(robot.id());
+        if (duplicate_id.second)
+        {
+            // TODO: Multiple robots on the same team with the same id. Throw exception
+            // See https://github.com/UBC-Thunderbots/Software/issues/16
+        }
+
+        auto it = team_robots.find(robot.id());
+        if (it != team_robots.end())
+        {
+            // The robot already exists on the team. Find and update the robot
+            team_robots.at(robot.id()).updateState(robot);
+        }
+        else
+        {
+            // This robot does not exist as part of the team yet. Add the new robot
+            team_robots.insert(std::make_pair(robot.id(), robot));
+        }
+    }
+
+    removeExpiredRobots(timestamp);
+}
+
+void Team::updateState(const Team& new_team_data,
+                       const std::chrono::steady_clock::time_point timestamp)
+{
+    updateRobots(new_team_data.getAllRobots(), timestamp);
+    this->goalie_id = new_team_data.goalie_id;
+}
+
+void Team::updateStateToPredictedState(
+    const std::chrono::steady_clock::time_point timestamp)
+{
+    // Update the state of all robots to their predicted state
+    for (auto it = team_robots.begin(); it != team_robots.end(); it++)
+    {
+        it->second.updateStateToPredictedState(timestamp);
+    }
 }
 
-void Team::update(const thunderbots_msgs::Team &team_msg)
+void Team::removeExpiredRobots(const std::chrono::steady_clock::time_point timestamp)
 {
+    // Check to see if any Robots have "expired". If it more time than the expiry_buffer
+    // has passed, then remove the robot from the team
+    for (auto it = team_robots.begin(); it != team_robots.end();)
+    {
+        if (std::chrono::duration(timestamp - it->second.lastUpdateTimestamp()) >
+            robot_expiry_buffer_milliseconds)
+        {
+            team_robots.erase(it++);
+        }
+        else
+        {
+            it++;
+        }
+    }
+}
+
+void Team::assignGoalie(unsigned int new_goalie_id)
+{
+    goalie_id = new_goalie_id;
+}
+
+void Team::clearGoalie()
+{
+    goalie_id.reset();
 }
 
 std::size_t Team::size() const
 {
-    return static_cast<size_t>(0);
+    return team_robots.size();
+}
+
+std::chrono::milliseconds Team::getRobotExpiryBufferMilliseconds()
+{
+    return robot_expiry_buffer_milliseconds;
 }
 
 std::optional<Robot> Team::getRobotById(unsigned int id) const
 {
+    auto it = team_robots.find(id);
+    if (it != team_robots.end())
+    {
+        return it->second;
+    }
+
     return std::nullopt;
 }
 
 std::optional<Robot> Team::goalie() const
 {
+    if (goalie_id)
+    {
+        return getRobotById(*goalie_id);
+    }
+
     return std::nullopt;
 }
 
 std::vector<Robot> Team::getAllRobots() const
 {
-    return std::vector<Robot>();
-}
\ No newline at end of file
+    std::vector<Robot> all_robots;
+    for (auto it = team_robots.begin(); it != team_robots.end(); it++)
+    {
+        all_robots.emplace_back(it->second);
+    }
```

Comment:
Another option here could be something like:
```
    std::vector<int> to_vector;
    std::copy(from_vector.begin(), from_vector.end(),
              std::back_inserter(to_vector));
```
Even if we don't go with the above, we should at least change this to something like:
```
for (auto robot : team_robots) 
{
    all_robots.emplace_back(robot);
}
```
This is true for many cases where we're manually iterating over something.

Reply:
team_robots is an `std::map`, so this type of iteration doesn't work.